### ISSUE TITLE
[Feature] GAN 学習方式を PLCS・trajectory_completion・event_detection に拡張

### ISSUE THREAD

**Issue by Motoki0705** [Motoki0705/tennis-lab#382]
## Summary

BLCS のみで実装されている GAN 学習方式を PLCS・trajectory_completion・event_detection にも拡張する。

## Problem

現在 GAN を用いた学習（`gan_training_strategy.py` 等）は `src/tasks/blcs/training/` にのみ存在する。他の 3 タスクでも GAN を活用できる余地があるが、対応する実装がない。

## Proposal

- PLCS・trajectory_completion・event_detection の各タスクに GAN 学習ストラテジーおよび関連コンポーネントを追加する
- BLCS の実装を参考に、各タスクのモデル・損失関数に合わせた discriminator と GAN ロスを導入する

## Scope

- 対象タスク: PLCS, trajectory_completion, event_detection（BLCS は既存実装を参照）

## Acceptance Criteria

- 各タスクで GAN 学習を有効化できる設定・コードが存在する
- GAN フェーズに移行した際に discriminator の学習が正常に動作する

## Notes

-

**Comment by Motoki0705** [comment]
実装前の設計メモです。承認をもらえたらこの方針の範囲で進めます。

## Requirements
- BLCS にある GAN 学習方式を PLCS・trajectory_completion・event_detection に拡張する
- 各タスクで設定から GAN 学習を有効化できるようにする
- GAN フェーズ移行後に discriminator 学習が正常に走るようにする

## Scope
- PLCS: `src/tasks/plcs/training/runner.py`, `src/tasks/plcs/training/lightning_module.py`, GAN 用 config, discriminator, smoke test 追加
- trajectory_completion: `src/tasks/trajectory_completion/training/runner.py`, `src/tasks/trajectory_completion/training/lightning_module.py`, GAN 用 config, discriminator, smoke test 追加
- event_detection: `src/tasks/event_detection/training/runner.py`, `src/tasks/event_detection/training/lightning_module.py`, GAN 用 config, discriminator, smoke test 追加
- 参照実装: `src/tasks/blcs/training/gan_training_strategy.py`, `src/tasks/blcs/training/gan_transition_callback.py`, `src/tasks/blcs/training/lightning_module.py`

## Proposed Approach
- まず BLCS の GAN 制御ロジックを各タスクへ持ち込み、runner 側で GAN 有効時の runtime config と transition callback を追加します
- discriminator はタスクごとに別実装にします。PLCS と trajectory_completion は軌跡系列を入力にした discriminator、event_detection はイベント logits 系列を入力にした discriminator にして、BLCS 用 discriminator をそのまま共通化するのはこの issue では行いません
- `training/default.yaml` に `optional gan: null` を追加し、各タスクに `training/gan/lsgan.yaml` を追加して Hydra override で有効化できるようにします
- `tests/tasks/test_training_smoke_contracts.py` に各タスクの GAN variant を追加し、GAN フェーズ移行まで含めた smoke を回します

## Acceptance Criteria
- `training/gan=lsgan` を付けた各対象タスクの train config が解決できる
- LightningModule が generator/discriminator の 2 optimizer 構成で GAN フェーズへ移行できる
- smoke test で各対象タスクの GAN variant が通る

## Risks / Notes
- event_detection は回帰系列ではなく `(B, T, E)` logits を扱うため、discriminator 入力と generator loss の取り方は専用化します
- 共通 GAN 基盤への切り出しは別 issue に分け、今回はタスク追加に絞ります

## Intended Branch
- `feat/issue-382-extend-gan-training`

**Comment by Motoki0705** [comment]
Implementation summary for feat/issue-382-extend-gan-training

Changed files
- Shared GAN utilities added under src/tasks/base/training/: LSGAN loss, generic GAN transition callback, and reusable manual GAN training support.
- Shared transformer discriminator added at src/utils/models/transformer_sequence_discriminator.py.
- PLCS / trajectory_completion / event_detection each now have:
  - task-specific discriminator factory + discriminator module
  - training defaults extended with training/_gan.yaml
  - runner-side GAN runtime handling and transition callback registration
  - LightningModule-side manual GAN optimization wiring
- BLCS gan_transition_callback.py now re-exports the shared callback for backward compatibility.
- tests/tasks/test_training_smoke_contracts.py now includes GAN smoke variants for PLCS, trajectory_completion, and event_detection.

Validation performed
- pytest tests/tasks/test_training_smoke_contracts.py -k 'plcs_chunked_multiview_gan' -vv -x
- pytest tests/tasks/test_training_smoke_contracts.py -k 'trajectory_completion_chunked_uv_transformer_gan or event_detection_uv_chunked_uv_transformer_gan or event_detection_3d_chunked_traj3d_transformer_gan' -vv -x
- pytest tests/tasks/test_training_smoke_contracts.py -k 'gan' -vv -x

Results
- All GAN smoke variants passed, including existing BLCS GAN coverage and the new PLCS / trajectory_completion / event_detection variants.
- The smoke assertions confirmed GAN phase transition, discriminator participation, and hybrid step execution.

Unresolved risks
- Validation is currently smoke-level only; no long-horizon training quality check was run.
- Non-chunked GAN paths were enabled in config/code but not exercised separately because the chunked smoke suite already covers the transition/callback/manual-optimization contract.

Completion criteria
- Each target task now has config and code paths to enable GAN training.
- GAN transition and discriminator updates are exercised and passing in smoke validation.
- Acceptance criteria are met for the implemented scope.

**Comment by Motoki0705** [comment]
Review revision plan

Requested direction
- Replace the three BLCS GAN training files with the shared base GAN implementation so BLCS / PLCS / trajectory_completion / event_detection use one consistent training contract.
- Replace BLCS trajectory discriminator with the generic transformer-sequence discriminator path.
- Move TransformerSequenceDiscriminator under src/utils/models/architectures.

Revised implementation scope
- Standardize BLCS LightningModule GAN result payload to the same keys already used by PLCS / trajectory_completion / event_detection: gan_fake, gan_real, gan_mask.
- Switch BLCS LightningModule from BLCSGANTrainingStrategy + local LSGANLoss to base ManualGANSupportMixin / ManualGANTrainingStrategy / LSGANLoss.
- Remove BLCS-local GAN duplication where it becomes redundant, keeping only compatibility re-exports if needed by current imports.
- Move src/utils/models/transformer_sequence_discriminator.py to src/utils/models/architectures/transformer_sequence_discriminator.py and update exports/imports.
- Convert src/tasks/blcs/models/discriminators/trajectory_discriminator.py into a thin BLCS wrapper/subclass over the moved generic discriminator so BLCS keeps its existing factory/config surface while using the shared implementation.
- Re-run BLCS GAN smoke plus all GAN smoke variants to confirm the unified path still works.

Acceptance criteria for this revision
- BLCS no longer owns a separate GAN training strategy / GAN loss implementation in practice.
- BLCS discriminator is implemented on top of the shared transformer-sequence discriminator.
- Shared discriminator import path is updated to the architectures package everywhere.
- Existing GAN smoke variants, including BLCS ones, still pass.

Notes
- The one intentional BLCS-specific residue may be a thin compatibility wrapper/re-export layer if that is the smallest safe way to preserve current imports and factory names.

**Comment by Motoki0705** [comment]
Revision result summary for feat/issue-382-extend-gan-training

Completed changes
- BLCS LightningModule now follows the same shared GAN contract as PLCS / trajectory_completion / event_detection by using the base manual GAN support path and returning gan_fake / gan_real / gan_mask from supervised results.
- The three BLCS GAN files now defer to base implementations in practice:
  - src/tasks/blcs/training/gan_loss.py -> shared LSGANLoss re-export
  - src/tasks/blcs/training/gan_training_strategy.py -> shared ManualGANTrainingStrategy alias
  - src/tasks/blcs/training/gan_transition_callback.py -> shared callback re-export
- Shared discriminator implementation moved from src/utils/models/transformer_sequence_discriminator.py to src/utils/models/architectures/transformer_sequence_discriminator.py.
- All task-specific discriminator wrappers now import from src/utils/models/architectures.
- src/tasks/blcs/models/discriminators/trajectory_discriminator.py is now a thin BLCS wrapper over the shared TransformerSequenceDiscriminator while preserving the BLCS config/factory surface.

Validation performed
- pytest tests/tasks/test_training_smoke_contracts.py -k 'blcs_single_gan or blcs_chunked_multiview_gan' -vv -x
- pytest tests/tasks/test_training_smoke_contracts.py -k 'gan' -vv -x

Results
- BLCS GAN smoke passed after the refactor.
- All GAN smoke variants passed across BLCS / PLCS / trajectory_completion / event_detection after the path move and shared discriminator update.

Residual risks
- BLCS discriminator now uses the shared generic input projection path rather than the previous Ball3DEmbedding-specific implementation, so any quality difference would only show up in longer training runs and is not covered by smoke tests.
- Compatibility shims were kept for BLCS-local GAN file paths so current imports remain stable while implementation is unified.

Revision acceptance
- BLCS no longer uses a distinct GAN training implementation path.
- BLCS discriminator is implemented on top of the shared transformer-sequence discriminator.
- Shared discriminator path is updated under architectures.
- GAN smoke coverage remains green.

**Comment by Motoki0705** [comment]
Review revision plan

Requested direction
- Remove the three BLCS-local GAN re-export files instead of keeping compatibility shims.
- Make the base lightning/runner layer thicker so downstream GAN tasks carry less repeated implementation.
- Keep the refactor validated by smoke tests.

Requirements
- Delete the redundant BLCS GAN shim files under src/tasks/blcs/training.
- Update BLCS imports so it uses base GAN symbols directly.
- Refactor base training helpers so GAN-enabled downstream lightning modules and runners need less repeated code.
- Keep behavior unchanged for the existing GAN-enabled tasks other than the intended code-path consolidation.
- Re-run the affected smoke contracts and confirm green.

Planned scope
1. Remove src/tasks/blcs/training/gan_loss.py, src/tasks/blcs/training/gan_training_strategy.py, and src/tasks/blcs/training/gan_transition_callback.py, then redirect any remaining references to src/tasks/base/training.
2. Expand the base lightning support layer to own the repeated GAN/manual-step stage dispatch and shared epoch-end logging hooks where the existing task modules already follow the same contract.
3. Expand the base training runner to own the repeated GAN runtime mutation and GAN transition callback wiring, leaving task runners to provide only task-specific callbacks/datamodule differences.
4. Trim BLCS runner/lightning_module to use the thicker base hooks, and apply the same base hook adoption to the other GAN-enabled task runners/lightning modules where it reduces duplication without changing task-specific loss/metric logic.
5. Re-run BLCS-focused smoke plus the full GAN smoke slice.

Acceptance criteria
- The three BLCS GAN shim files are removed.
- BLCS imports GAN support directly from src/tasks/base/training.
- Base lightning/runner own more of the shared GAN path, and downstream task code is correspondingly smaller.
- BLCS smoke and the full GAN smoke slice both pass after the refactor.

Working branch
- feat/issue-382-extend-gan-training

**Comment by Motoki0705** [comment]
Revision result summary for feat/issue-382-extend-gan-training

Completed changes
- Removed the BLCS-local GAN shim files:
  - src/tasks/blcs/training/gan_loss.py
  - src/tasks/blcs/training/gan_training_strategy.py
  - src/tasks/blcs/training/gan_transition_callback.py
- BLCS now imports and uses base GAN support directly, and src/tasks/blcs/training/__init__.py no longer re-exports base GAN symbols.
- Thickened src/tasks/base/training/gan_training.py so ManualGANSupportMixin now owns the repeated train/val/test stage dispatch and shared epoch-end metric flushing hooks for GAN-capable modules.
- Thickened src/tasks/base/training/runner.py so BaseTrainingRunner now owns the shared GAN runtime config mutation and GANTransitionCallback injection.
- BLCS / PLCS / event_detection / trajectory_completion runners now only add task-specific preparation and chunk-rotation behavior on top of the shared base GAN runner hooks.
- BLCS / PLCS / event_detection / trajectory_completion lightning modules now keep task-specific supervised result + logging logic while delegating the repeated GAN stage flow to the base mixin.

Validation performed
- /home/kamimura/projects/tennis-lab/.venv/bin/python -m pytest tests/tasks/test_training_smoke_contracts.py -k 'blcs_single_gan or blcs_chunked_multiview_gan' -vv -x
- /home/kamimura/projects/tennis-lab/.venv/bin/python -m pytest tests/tasks/test_training_smoke_contracts.py -k 'gan' -vv -x

Results
- BLCS GAN smoke passed after removing the BLCS-local GAN shims.
- Full GAN smoke passed across BLCS / PLCS / trajectory_completion / event_detection after the base runner and base lightning refactor.

Residual risks
- The shared lightning hook layer assumes the existing GAN-enabled tasks continue to provide the same _compute_supervised_result and _log_stage_metrics contracts; if a future task deviates from that shape it will need its own override path.
- Validation remains smoke-level only; no longer-horizon training quality check was run.

Revision acceptance
- BLCS no longer carries the three local GAN re-export files.
- Base lightning/runner own more of the shared GAN path and downstream task code is smaller.
- BLCS-focused and full GAN smoke validation are both green.
